resolver/resolvers.go

```diff
@@ -92,65 +98,60 @@ func createResolver(resolverURL, source string) (*Resolver, bool, error) {
 	}
 
 	switch u.Scheme {
-	case ServerTypeDNS, ServerTypeDoT, ServerTypeTCP:
+	case ServerTypeDNS, ServerTypeDoT, ServerTypeDoH, ServerTypeTCP:
+	case HttpsProtocol:
+		u.Scheme = ServerTypeDoH
 	default:
 		return nil, false, fmt.Errorf("DNS resolver scheme %q invalid", u.Scheme)
 	}
 
 	ip := net.ParseIP(u.Hostname())
-	if ip == nil {
+	isHostnameDomain := (ip == nil)
+	if ip == nil && u.Scheme != ServerTypeDoH {
 		return nil, false, fmt.Errorf("resolver IP %q invalid", u.Hostname())
 	}
 
 	// Add default port for scheme if it is missing.
-	var port uint16
-	hostPort := u.Port()
-	switch {
-	case hostPort != "":
-		parsedPort, err := strconv.ParseUint(hostPort, 10, 16)
-		if err != nil {
-			return nil, false, fmt.Errorf("resolver port %q invalid", u.Port())
-		}
-		port = uint16(parsedPort)
-	case u.Scheme == ServerTypeDNS, u.Scheme == ServerTypeTCP:
-		port = 53
-	case u.Scheme == ServerTypeDoH:
-		port = 443
-	case u.Scheme == ServerTypeDoT:
-		port = 853
-	default:
-		return nil, false, fmt.Errorf("missing port in %q", u.Host)
-	}
-
-	scope := netutils.GetIPScope(ip)
-	// Skip localhost resolvers from the OS, but not if configured.
-	if scope.IsLocalhost() && source == ServerSourceOperatingSystem {
-		return nil, true, nil // skip
+	port, err := parsePortFromURL(u)
+	if err != nil {
+		return nil, false, err
 	}
 
 	// Get parameters and check if keys exist.
 	query := u.Query()
-	for key := range query {
-		switch key {
-		case parameterName,
-			parameterVerify,
-			parameterBlockedIf,
-			parameterSearch,
-			parameterSearchOnly:
-			// Known key, continue.
-		default:
-			// Unknown key, abort.
-			return nil, false, fmt.Errorf(`unknown parameter "%s"`, key)
-		}
+	err = checkURLParameterValidity(u.Scheme, isHostnameDomain, query)
+	if err != nil {
+		return nil, false, err
 	}
 
 	// Check domain verification config.
+	serverAddress := ""
+	paramterServerIP := query.Get(parameterIP)
```

Comment:
typo: `parameterServerIP`
suggest: `serverIPParameter` as to not confuse with the `parameter*` constants.

Also, the IP address should be validated.